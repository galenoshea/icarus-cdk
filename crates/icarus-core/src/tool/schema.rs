//! JSON schema types for tool parameters.
//!
//! Provides type-safe representation of JSON Schema for parameter validation.

use std::collections::HashMap;

use candid::{CandidType, Deserialize};
use serde::Serialize;

use crate::{IcarusError, ToolId};

/// JSON schema types for tool parameters.
///
/// Provides type-safe representation of JSON Schema for parameter validation.
#[derive(Debug, Clone, CandidType, Deserialize, Serialize)]
#[serde(tag = "type", rename_all = "lowercase")]
pub enum ToolSchema {
    /// String type with optional constraints.
    String {
        /// Minimum string length.
        #[serde(skip_serializing_if = "Option::is_none")]
        min_length: Option<usize>,
        /// Maximum string length.
        #[serde(skip_serializing_if = "Option::is_none")]
        max_length: Option<usize>,
        /// Pattern for validation (regex).
        #[serde(skip_serializing_if = "Option::is_none")]
        pattern: Option<String>,
        /// Enumerated values (if applicable).
        #[serde(skip_serializing_if = "Option::is_none")]
        r#enum: Option<Vec<String>>,
    },
    /// Number type (integer or float).
    Number {
        /// Minimum value.
        #[serde(skip_serializing_if = "Option::is_none")]
        minimum: Option<f64>,
        /// Maximum value.
        #[serde(skip_serializing_if = "Option::is_none")]
        maximum: Option<f64>,
    },
    /// Integer type.
    Integer {
        /// Minimum value.
        #[serde(skip_serializing_if = "Option::is_none")]
        minimum: Option<i64>,
        /// Maximum value.
        #[serde(skip_serializing_if = "Option::is_none")]
        maximum: Option<i64>,
    },
    /// Boolean type.
    Boolean,
    /// Array type with item schema.
    Array {
        /// Schema for array items.
        items: Box<ToolSchema>,
        /// Minimum number of items.
        #[serde(skip_serializing_if = "Option::is_none")]
        min_items: Option<usize>,
        /// Maximum number of items.
        #[serde(skip_serializing_if = "Option::is_none")]
        max_items: Option<usize>,
    },
    /// Object type with property schemas.
    Object {
        /// Schemas for object properties.
        properties: HashMap<String, ToolSchema>,
        /// Required property names.
        required: Vec<String>,
    },
}

impl ToolSchema {
    /// Creates a string schema.
    #[must_use]
    pub fn string() -> Self {
        Self::String {
            min_length: None,
            max_length: None,
            pattern: None,
            r#enum: None,
        }
    }

    /// Creates a string schema with length constraints.
    #[must_use]
    pub fn string_with_length(min: Option<usize>, max: Option<usize>) -> Self {
        Self::String {
            min_length: min,
            max_length: max,
            pattern: None,
            r#enum: None,
        }
    }

    /// Creates a string schema with enumerated values.
    #[must_use]
    pub fn string_enum(values: impl IntoIterator<Item = impl Into<String>>) -> Self {
        Self::String {
            min_length: None,
            max_length: None,
            pattern: None,
            r#enum: Some(values.into_iter().map(Into::into).collect()),
        }
    }

    /// Creates a number schema.
    #[must_use]
    pub fn number() -> Self {
        Self::Number {
            minimum: None,
            maximum: None,
        }
    }

    /// Creates a number schema with range constraints.
    #[must_use]
    pub fn number_range(min: Option<f64>, max: Option<f64>) -> Self {
        Self::Number {
            minimum: min,
            maximum: max,
        }
    }

    /// Creates an integer schema.
    #[must_use]
    pub fn integer() -> Self {
        Self::Integer {
            minimum: None,
            maximum: None,
        }
    }

    /// Creates an integer schema with range constraints.
    #[must_use]
    pub fn integer_range(min: Option<i64>, max: Option<i64>) -> Self {
        Self::Integer {
            minimum: min,
            maximum: max,
        }
    }

    /// Creates a boolean schema.
    #[must_use]
    pub fn boolean() -> Self {
        Self::Boolean
    }

    /// Creates an array schema.
    #[must_use]
    pub fn array(items: Self) -> Self {
        Self::Array {
            items: Box::new(items),
            min_items: None,
            max_items: None,
        }
    }

    /// Creates an object schema.
    #[must_use]
    pub fn object(
        properties: impl IntoIterator<Item = (impl Into<String>, Self)>,
        required: impl IntoIterator<Item = impl Into<String>>,
    ) -> Self {
        Self::Object {
            properties: properties.into_iter().map(|(k, v)| (k.into(), v)).collect(),
            required: required.into_iter().map(Into::into).collect(),
        }
    }

    /// Validates the schema definition.
    ///
    /// # Errors
    ///
    /// Returns `IcarusError::InvalidSchema` if the schema is invalid.
    pub fn validate(&self) -> Result<(), IcarusError> {
        match self {
            Self::String {
                min_length,
                max_length,
                ..
            } => {
                if let (Some(min), Some(max)) = (min_length, max_length) {
                    if min > max {
                        return Err(IcarusError::InvalidSchema {
                            tool_id: ToolId::new("unknown").unwrap_or_else(|_| unreachable!()),
                            message: "String min_length cannot be greater than max_length"
                                .to_string(),
                        });
                    }
                }
            }
            Self::Number { minimum, maximum } => {
                if let (Some(min), Some(max)) = (minimum, maximum) {
                    if min > max {
                        return Err(IcarusError::InvalidSchema {
                            tool_id: ToolId::new("unknown").unwrap_or_else(|_| unreachable!()),
                            message: "Number minimum cannot be greater than maximum".to_string(),
                        });
                    }
                }
            }
            Self::Integer { minimum, maximum } => {
                if let (Some(min), Some(max)) = (minimum, maximum) {
                    if min > max {
                        return Err(IcarusError::InvalidSchema {
                            tool_id: ToolId::new("unknown").unwrap_or_else(|_| unreachable!()),
                            message: "Integer minimum cannot be greater than maximum".to_string(),
                        });
                    }
                }
            }
            Self::Array {
                items,
                min_items,
                max_items,
            } => {
                items.validate()?;
                if let (Some(min), Some(max)) = (min_items, max_items) {
                    if min > max {
                        return Err(IcarusError::InvalidSchema {
                            tool_id: ToolId::new("unknown").unwrap_or_else(|_| unreachable!()),
                            message: "Array min_items cannot be greater than max_items".to_string(),
                        });
                    }
                }
            }
            Self::Object { properties, .. } => {
                for schema in properties.values() {
                    schema.validate()?;
                }
            }
            Self::Boolean => {}
        }

        Ok(())
    }
}
