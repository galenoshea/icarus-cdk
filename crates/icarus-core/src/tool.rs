//! Tool abstraction for Icarus MCP servers

use crate::error::Result;
use async_trait::async_trait;
use rmcp::model::Tool;
use serde::{Deserialize, Serialize};
use serde_json::Value;

/// Icon information for tools
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct IconInfo {
    /// Icon name or identifier
    pub name: String,
    /// Optional icon data (base64 encoded or URL)
    pub data: Option<String>,
}

/// Information about a tool
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ToolInfo {
    pub name: String,
    pub description: String,
    pub input_schema: Value,
    /// Optional display title for better UX
    pub title: Option<String>,
    /// Optional icons for tool display
    pub icons: Option<Vec<IconInfo>>,
}

/// Storage requirements for a tool
#[derive(Debug, Clone, Default)]
pub struct StorageRequirements {
    /// Estimated stable memory usage in bytes
    pub stable_memory: Option<u64>,
    /// Whether the tool needs persistent state
    pub requires_persistence: bool,
}

/// Core trait for Icarus tools that integrate with rmcp
#[async_trait]
pub trait IcarusTool: Send + Sync {
    /// Get tool information
    fn info(&self) -> ToolInfo;

    /// Convert to rmcp tool representation
    fn to_rmcp_tool(&self) -> Tool {
        use std::borrow::Cow;
        use std::sync::Arc;

        let info = self.info();

        // Convert input_schema Value to JsonObject
        let schema = if let serde_json::Value::Object(obj) = &info.input_schema {
            Arc::new(obj.clone())
        } else {
            Arc::new(serde_json::Map::new())
        };

        Tool {
            name: Cow::Owned(info.name),
            description: Some(Cow::Owned(info.description)),
            input_schema: schema,
            output_schema: None,
            annotations: None,
            title: info.title,
            icons: None, // Icons will be handled later when we support them fully
        }
    }

    /// Check if the tool requires stable storage
    fn requires_stable_storage(&self) -> bool {
        false
    }

    /// Get storage requirements for capacity planning
    fn storage_requirements(&self) -> StorageRequirements {
        StorageRequirements::default()
    }

    /// Execute the tool with given arguments
    async fn execute(&self, args: Value) -> Result<Value>;
}
