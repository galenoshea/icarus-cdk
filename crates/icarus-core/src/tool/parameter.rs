//! Tool parameter definitions and utilities.
//!
//! This module provides parameter-related types including the `SmallParameters`
//! optimization wrapper and `ToolParameter` struct.

use std::ops::{Deref, DerefMut};

use candid::{CandidType, Deserialize};
use serde::Serialize;
use smallvec::SmallVec;

use crate::{IcarusError, ToolId};

use super::schema::ToolSchema;

/// A newtype wrapper around `SmallVec` that implements `CandidType` for IC compatibility.
///
/// This wrapper provides the performance benefits of `SmallVec` for collections with ≤4 elements
/// while maintaining compatibility with Candid serialization for Internet Computer canisters.
/// The wrapper transparently converts to/from `Vec` during serialization.
///
/// # Performance
///
/// - Stack-allocated for ≤4 elements (zero heap allocations)
/// - Heap-allocated `Vec` fallback for larger collections
/// - Transparent access to all `SmallVec` methods via `Deref`/`DerefMut`
///
/// # Examples
///
/// ```rust
/// use icarus_core::tool::SmallParameters;
///
/// let mut params = SmallParameters::<i32>::new();
/// params.push(42);
/// params.push(100);
///
/// let params = SmallParameters::from_vec(vec![1, 2, 3]);
/// assert_eq!(params.len(), 3);
/// ```
#[derive(Debug, Clone)]
pub struct SmallParameters<T>(SmallVec<[T; 4]>);

impl<T> SmallParameters<T> {
    /// Creates a new empty `SmallParameters` collection.
    #[must_use]
    #[inline]
    pub fn new() -> Self {
        Self(SmallVec::new())
    }

    /// Creates a new `SmallParameters` collection with the specified capacity.
    #[must_use]
    #[inline]
    pub fn with_capacity(capacity: usize) -> Self {
        Self(SmallVec::with_capacity(capacity))
    }

    /// Creates a `SmallParameters` collection from a `Vec`.
    #[must_use]
    #[inline]
    pub fn from_vec(vec: Vec<T>) -> Self {
        Self(SmallVec::from_vec(vec))
    }

    /// Converts the `SmallParameters` to a `Vec`.
    #[must_use]
    #[inline]
    pub fn into_vec(self) -> Vec<T> {
        self.0.into_vec()
    }

    /// Returns a reference to the inner `SmallVec`.
    #[must_use]
    #[inline]
    pub fn as_smallvec(&self) -> &SmallVec<[T; 4]> {
        &self.0
    }

    /// Returns a mutable reference to the inner `SmallVec`.
    #[must_use]
    #[inline]
    pub fn as_smallvec_mut(&mut self) -> &mut SmallVec<[T; 4]> {
        &mut self.0
    }
}

impl<T> Default for SmallParameters<T> {
    #[inline]
    fn default() -> Self {
        Self::new()
    }
}

impl<T> Deref for SmallParameters<T> {
    type Target = SmallVec<[T; 4]>;

    #[inline]
    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl<T> DerefMut for SmallParameters<T> {
    #[inline]
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.0
    }
}

impl<T> From<Vec<T>> for SmallParameters<T> {
    #[inline]
    fn from(vec: Vec<T>) -> Self {
        Self::from_vec(vec)
    }
}

impl<T> From<SmallVec<[T; 4]>> for SmallParameters<T> {
    #[inline]
    fn from(small_vec: SmallVec<[T; 4]>) -> Self {
        Self(small_vec)
    }
}

impl<T> From<SmallParameters<T>> for Vec<T> {
    #[inline]
    fn from(small_params: SmallParameters<T>) -> Self {
        small_params.into_vec()
    }
}

impl<T> FromIterator<T> for SmallParameters<T> {
    #[inline]
    fn from_iter<I: IntoIterator<Item = T>>(iter: I) -> Self {
        Self(SmallVec::from_iter(iter))
    }
}

impl<T> Extend<T> for SmallParameters<T> {
    #[inline]
    fn extend<I: IntoIterator<Item = T>>(&mut self, iter: I) {
        self.0.extend(iter);
    }
}

// Candid serialization: convert to/from Vec for IC compatibility
impl<T: CandidType + Clone> CandidType for SmallParameters<T> {
    fn _ty() -> candid::types::Type {
        Vec::<T>::_ty()
    }

    fn idl_serialize<S>(&self, serializer: S) -> Result<(), S::Error>
    where
        S: candid::types::Serializer,
    {
        // Convert to Vec for serialization
        let vec: Vec<T> = self.0.iter().cloned().collect();
        vec.idl_serialize(serializer)
    }
}

// Serde serialization: convert to/from Vec
impl<T: Serialize> Serialize for SmallParameters<T> {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        self.0.serialize(serializer)
    }
}

impl<'de, T: Deserialize<'de>> Deserialize<'de> for SmallParameters<T> {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        let vec = Vec::<T>::deserialize(deserializer)?;
        Ok(Self::from_vec(vec))
    }
}

/// Tool parameter definition with schema information.
///
/// Represents a single parameter that a tool accepts, including its type,
/// validation rules, and documentation.
#[derive(Debug, Clone, CandidType, Deserialize, Serialize)]
pub struct ToolParameter {
    /// Parameter name (must be a valid identifier).
    pub name: String,
    /// Human-readable description of the parameter.
    pub description: String,
    /// JSON schema for the parameter type and validation.
    pub schema: ToolSchema,
    /// Whether this parameter is required.
    pub required: bool,
    /// Default value for optional parameters as JSON string.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub default: Option<String>,
}

impl ToolParameter {
    /// Creates a new required parameter.
    #[must_use]
    pub fn new(
        name: impl Into<String>,
        description: impl Into<String>,
        schema: ToolSchema,
    ) -> Self {
        Self {
            name: name.into(),
            description: description.into(),
            schema,
            required: true,
            default: None,
        }
    }

    /// Creates a new optional parameter.
    #[must_use]
    pub fn optional(
        name: impl Into<String>,
        description: impl Into<String>,
        schema: ToolSchema,
    ) -> Self {
        Self {
            name: name.into(),
            description: description.into(),
            schema,
            required: false,
            default: None,
        }
    }

    /// Creates a new optional parameter with a default value.
    #[must_use]
    pub fn with_default(
        name: impl Into<String>,
        description: impl Into<String>,
        schema: ToolSchema,
        default: impl Into<String>,
    ) -> Self {
        Self {
            name: name.into(),
            description: description.into(),
            schema,
            required: false,
            default: Some(default.into()),
        }
    }

    /// Validates the parameter definition.
    ///
    /// # Errors
    ///
    /// Returns `IcarusError::InvalidSchema` if the parameter is invalid.
    pub fn validate(&self) -> Result<(), IcarusError> {
        if self.name.is_empty() {
            return Err(IcarusError::InvalidSchema {
                tool_id: ToolId::new("unknown").unwrap_or_else(|_| unreachable!()),
                message: "Parameter name cannot be empty".to_string(),
            });
        }

        if !self.name.chars().all(|c| c.is_alphanumeric() || c == '_') {
            return Err(IcarusError::InvalidSchema {
                tool_id: ToolId::new("unknown").unwrap_or_else(|_| unreachable!()),
                message: format!("Invalid parameter name: {}", self.name),
            });
        }

        if self.description.is_empty() {
            return Err(IcarusError::InvalidSchema {
                tool_id: ToolId::new("unknown").unwrap_or_else(|_| unreachable!()),
                message: format!("Parameter '{}' description cannot be empty", self.name),
            });
        }

        self.schema.validate()?;

        Ok(())
    }
}
