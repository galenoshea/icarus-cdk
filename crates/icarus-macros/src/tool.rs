//! Implementation of the #[tool] attribute macro.

use proc_macro2::TokenStream;
use quote::{format_ident, quote};
use syn::{parse2, spanned::Spanned, ItemFn};

use crate::error::{MacroError, MacroResult};
use crate::utils::{
    extract_parameters, extract_return_type, generate_function_call,
    generate_json_schema_from_parameters, generate_param_struct_name, is_async_function,
};

/// Maximum number of parameters a tool function can have
const MAX_PARAMETERS: usize = 50;

/// Implementation of the #[tool] attribute macro.
pub(crate) fn tool_impl(args: TokenStream, input: TokenStream) -> MacroResult<TokenStream> {
    // Parse the function
    let function: ItemFn = parse2(input)?;

    // Parse tool configuration from macro arguments
    let tool_config = if args.is_empty() {
        ToolConfig::default()
    } else {
        parse_tool_args(args)
    };

    // Validate the function signature
    validate_function_signature(&function)?;

    // Extract function information
    let fn_name = &function.sig.ident;
    let fn_vis = &function.vis;
    let fn_attrs = &function.attrs;
    let fn_sig = &function.sig;
    let fn_block = &function.block;
    let is_async = is_async_function(fn_sig);

    // Extract parameters and return type
    let parameters = extract_parameters(&function.sig.inputs)?;

    // Validate parameter count to prevent pathological cases
    if parameters.len() > MAX_PARAMETERS {
        return Err(MacroError::invalid_signature_spanned(
            format!(
                "Tool functions cannot have more than {} parameters (found {}). \
                 Large parameter counts lead to slow compilation and large generated code.",
                MAX_PARAMETERS,
                parameters.len()
            ),
            function.sig.span(),
        ));
    }

    let _return_type = extract_return_type(&function.sig.output);

    // Generate parameter structure
    let param_struct_name = generate_param_struct_name(fn_name);
    let param_struct = generate_parameter_struct(&param_struct_name, &parameters);

    // Generate tool wrapper function
    let wrapper_fn_name = format_ident!("{}_tool_wrapper", fn_name);
    let tool_wrapper = generate_tool_wrapper(
        &wrapper_fn_name,
        fn_name,
        &param_struct_name,
        &parameters,
        is_async,
        tool_config.auth_level.as_deref(),
    );

    // Generate tool registration
    let registration_fn_name = format_ident!("{}_tool_info", fn_name);
    let description = tool_config
        .description
        .or_else(|| extract_doc_comment(fn_attrs));

    // Determine the tool name (custom or default)
    let default_tool_name = fn_name.to_string();
    let tool_name = tool_config.name.as_deref().unwrap_or(&default_tool_name);

    let tool_registration = generate_tool_info_function(
        &registration_fn_name,
        tool_name,
        &parameters,
        description.as_deref(),
        tool_config.auth_level.as_deref(),
    );

    // Generate linkme registration for automatic tool discovery
    let tool_registry_item = generate_tool_registry_item(&registration_fn_name);

    // Generate executor registration for runtime tool execution
    let executor_registration =
        generate_executor_registration(tool_name, &wrapper_fn_name, is_async);

    // Keep the original function unchanged
    let original_function = quote! {
        #(#fn_attrs)*
        #fn_vis #fn_sig #fn_block
    };

    // Combine all generated code
    Ok(quote! {
        #original_function

        #param_struct

        #tool_wrapper

        #tool_registration

        #tool_registry_item

        #executor_registration
    })
}

/// Configuration options for the #[tool] attribute.
#[derive(Debug, Default)]
struct ToolConfig {
    /// Optional custom tool name (allows kebab-case names for MCP compatibility)
    name: Option<String>,
    /// Optional custom description
    description: Option<String>,
    /// Authentication level: "none", "user", or "admin"
    auth_level: Option<String>,
}

/// Parses tool attribute arguments.
fn parse_tool_args(args: TokenStream) -> ToolConfig {
    use syn::parse::{Parse, ParseStream};
    use syn::Token;

    struct ToolArgs {
        name: Option<String>,
        description: Option<String>,
        auth_level: Option<String>,
    }

    impl Parse for ToolArgs {
        fn parse(input: ParseStream) -> syn::Result<Self> {
            let mut name = None;
            let mut description = None;
            let mut auth_level = None;

            // Try to parse the first argument as a string literal (description)
            if input.peek(syn::LitStr) {
                let lit: syn::LitStr = input.parse()?;
                description = Some(lit.value());

                // Parse remaining comma-separated arguments
                while !input.is_empty() {
                    let _: Token![,] = input.parse()?;

                    if input.is_empty() {
                        break;
                    }

                    let ident: syn::Ident = input.parse()?;
                    let _: Token![=] = input.parse()?;
                    let value: syn::LitStr = input.parse()?;

                    if ident == "auth" {
                        auth_level = Some(value.value());
                    } else if ident == "name" {
                        name = Some(value.value());
                    }
                }
            } else if input.peek(syn::Ident) {
                // Parse key=value pairs when no positional description
                while !input.is_empty() {
                    let ident: syn::Ident = input.parse()?;
                    let _: Token![=] = input.parse()?;
                    let value: syn::LitStr = input.parse()?;

                    if ident == "name" {
                        name = Some(value.value());
                    } else if ident == "description" {
                        description = Some(value.value());
                    } else if ident == "auth" {
                        auth_level = Some(value.value());
                    }

                    // Check for trailing comma
                    if input.peek(Token![,]) {
                        let _: Token![,] = input.parse()?;
                    } else {
                        break;
                    }
                }
            }

            Ok(ToolArgs {
                name,
                description,
                auth_level,
            })
        }
    }

    let parsed = parse2::<ToolArgs>(args).unwrap_or(ToolArgs {
        name: None,
        description: None,
        auth_level: None,
    });

    ToolConfig {
        name: parsed.name,
        description: parsed.description,
        auth_level: parsed.auth_level,
    }
}

/// Validates that the function signature is suitable for a tool.
fn validate_function_signature(function: &ItemFn) -> MacroResult<()> {
    // Check for generic parameters
    if !function.sig.generics.params.is_empty() {
        return Err(MacroError::unsupported_feature_spanned(
            "Generic functions",
            "Tool functions cannot have generic parameters",
            function.sig.generics.span(),
        ));
    }

    // Check for lifetimes
    if function.sig.generics.lifetimes().count() > 0 {
        return Err(MacroError::unsupported_feature_spanned(
            "Lifetime parameters",
            "Tool functions cannot have lifetime parameters",
            function.sig.generics.span(),
        ));
    }

    // Check for self parameter
    for input in &function.sig.inputs {
        if let syn::FnArg::Receiver(receiver) = input {
            return Err(MacroError::invalid_signature_spanned(
                "Tool functions cannot have self parameters",
                receiver.span(),
            ));
        }
    }

    Ok(())
}

/// Generates a parameter structure for the tool.
fn generate_parameter_struct(
    struct_name: &syn::Ident,
    parameters: &[crate::utils::ParameterInfo],
) -> TokenStream {
    let field_definitions: Vec<TokenStream> = parameters
        .iter()
        .map(|param| {
            let name = &param.name;
            let ty = &param.ty;

            quote! {
                pub #name: #ty,
            }
        })
        .collect();

    quote! {
        #[derive(serde::Deserialize)]
        struct #struct_name {
            #(#field_definitions)*
        }
    }
}

/// Generates the tool wrapper function that handles MCP protocol.
///
/// This function generates two wrappers:
/// 1. Internal wrapper that returns `Result<String, String>`
/// 2. Executor wrapper that converts to `RuntimeResult<LegacyToolResult<'static>>`
#[allow(clippy::too_many_arguments)]
fn generate_tool_wrapper(
    wrapper_name: &syn::Ident,
    fn_name: &syn::Ident,
    param_struct_name: &syn::Ident,
    parameters: &[crate::utils::ParameterInfo],
    is_async: bool,
    auth_level: Option<&str>,
) -> TokenStream {
    let fn_call = generate_function_call(fn_name, parameters, is_async);

    // Generate auth check code if auth_level is specified
    let auth_check = match auth_level {
        Some("user") => quote! {
            {
                let caller = ::ic_cdk::caller();
                if !::icarus_core::auth::has_user_access(&caller) {
                    return Err("Authentication required: user or admin access needed".to_string());
                }
            }
        },
        Some("admin") => quote! {
            {
                let caller = ::ic_cdk::caller();
                if !::icarus_core::auth::has_admin_access(&caller) {
                    return Err("Authentication required: admin access needed".to_string());
                }
            }
        },
        _ => quote! {}, // "none" or no auth - no check needed
    };

    if is_async {
        quote! {
            async fn #wrapper_name(args_json: &str) -> Result<String, String> {
                #auth_check

                let args: #param_struct_name = serde_json::from_str(args_json)
                    .map_err(|e| format!("Invalid arguments: {e}"))?;

                let result = #fn_call;

                serde_json::to_string(&result)
                    .map_err(|e| format!("Failed to serialize result: {e}"))
            }
        }
    } else {
        quote! {
            fn #wrapper_name(args_json: &str) -> Result<String, String> {
                #auth_check

                let args: #param_struct_name = serde_json::from_str(args_json)
                    .map_err(|e| format!("Invalid arguments: {e}"))?;

                let result = #fn_call;

                serde_json::to_string(&result)
                    .map_err(|e| format!("Failed to serialize result: {e}"))
            }
        }
    }
}

/// Generates the tool information function for registration.
fn generate_tool_info_function(
    info_fn_name: &syn::Ident,
    tool_name: &str,
    parameters: &[crate::utils::ParameterInfo],
    description: Option<&str>,
    auth_level: Option<&str>,
) -> TokenStream {
    let default_description = format!("Tool: {tool_name}");
    let description = description.unwrap_or(&default_description);

    // Generate JSON Schema for input parameters
    let input_schema = generate_json_schema_from_parameters(parameters);

    // Generate annotations if auth_level is specified
    let annotations_code = if let Some(auth) = auth_level {
        // Map auth_level to RMCP ToolAnnotations hints
        let read_only = auth == "none"; // Public tools might be read-only

        quote! {
            let annotations = ::icarus_core::ToolAnnotations {
                title: None,
                read_only_hint: Some(#read_only),
                destructive_hint: None,
                idempotent_hint: None,
                open_world_hint: None,
            };
            tool = tool.annotate(annotations);
        }
    } else {
        quote! {}
    };

    quote! {
        fn #info_fn_name() -> ::icarus_core::Tool {
            let input_schema = #input_schema;

            let mut tool = ::icarus_core::Tool::new(
                #tool_name,
                #description,
                input_schema,
            );

            #annotations_code

            tool
        }
    }
}

/// Generates linkme registration for automatic tool discovery.
fn generate_tool_registry_item(info_fn_name: &syn::Ident) -> TokenStream {
    let registry_static_name =
        format_ident!("TOOL_{}_REGISTRY", info_fn_name.to_string().to_uppercase());

    quote! {
        #[::linkme::distributed_slice(::icarus_runtime::TOOL_REGISTRY)]
        static #registry_static_name: fn() -> ::icarus_core::Tool = #info_fn_name;
    }
}

/// Generates executor wrapper and registration for runtime tool execution.
///
/// This creates:
/// 1. An executor function that wraps the internal wrapper and converts results
/// 2. A registration function that registers the executor with the runtime
fn generate_executor_registration(
    tool_name: &str,
    wrapper_fn_name: &syn::Ident,
    is_async: bool,
) -> TokenStream {
    // Use the wrapper function name to derive executor names to avoid conflicts
    let executor_fn_name = format_ident!("{}_executor", wrapper_fn_name);
    // Generate UPPER_CASE static variable name for Rust conventions
    let registration_fn_name = format_ident!(
        "{}_REGISTRATION",
        wrapper_fn_name.to_string().to_uppercase()
    );

    if is_async {
        quote! {
            fn #executor_fn_name(args: &str) -> ::std::pin::Pin<::std::boxed::Box<dyn ::std::future::Future<Output = ::icarus_runtime::RuntimeResult<::icarus_core::LegacyToolResult<'static>>> + Send>> {
                let args = args.to_string();
                ::std::boxed::Box::pin(async move {
                    match #wrapper_fn_name(&args).await {
                        Ok(result_json) => {
                            Ok(::icarus_core::LegacyToolResult::success(::std::borrow::Cow::Owned(result_json)))
                        }
                        Err(error_msg) => {
                            Ok(::icarus_core::LegacyToolResult::error(::std::borrow::Cow::Owned(error_msg)))
                        }
                    }
                })
            }

            #[::linkme::distributed_slice(::icarus_runtime::EXECUTOR_INIT)]
            static #registration_fn_name: fn() = || {
                let tool_id = ::icarus_core::ToolId::new(#tool_name)
                    .unwrap_or_else(|e| {
                        ::std::panic!(
                            "Invalid tool name '{}': {}. Tool names must be valid identifiers.",
                            #tool_name, e
                        )
                    });

                let _ = ::icarus_runtime::ToolRegistry::register_async_executor(
                    tool_id,
                    #executor_fn_name
                );
            };
        }
    } else {
        quote! {
            fn #executor_fn_name(args: &str) -> ::icarus_runtime::RuntimeResult<::icarus_core::LegacyToolResult<'static>> {
                match #wrapper_fn_name(args) {
                    Ok(result_json) => {
                        Ok(::icarus_core::LegacyToolResult::success(::std::borrow::Cow::Owned(result_json)))
                    }
                    Err(error_msg) => {
                        Ok(::icarus_core::LegacyToolResult::error(::std::borrow::Cow::Owned(error_msg)))
                    }
                }
            }

            #[::linkme::distributed_slice(::icarus_runtime::EXECUTOR_INIT)]
            static #registration_fn_name: fn() = || {
                let tool_id = ::icarus_core::ToolId::new(#tool_name)
                    .unwrap_or_else(|e| {
                        ::std::panic!(
                            "Invalid tool name '{}': {}. Tool names must be valid identifiers.",
                            #tool_name, e
                        )
                    });

                let _ = ::icarus_runtime::ToolRegistry::register_sync_executor(
                    tool_id,
                    #executor_fn_name
                );
            };
        }
    }
}

/// Extracts documentation comment from function attributes.
fn extract_doc_comment(attrs: &[syn::Attribute]) -> Option<String> {
    let mut doc_parts = Vec::new();

    for attr in attrs {
        if attr.path().is_ident("doc") {
            if let syn::Meta::NameValue(meta) = &attr.meta {
                if let syn::Expr::Lit(syn::ExprLit {
                    lit: syn::Lit::Str(lit_str),
                    ..
                }) = &meta.value
                {
                    let content = lit_str.value();
                    doc_parts.push(content.trim().to_string());
                }
            }
        }
    }

    if doc_parts.is_empty() {
        None
    } else {
        Some(doc_parts.join(" "))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_extract_doc_comment() {
        let input: ItemFn = syn::parse_quote! {
            /// This is a test function
            /// that does something useful.
            fn test_fn() {}
        };

        let doc = extract_doc_comment(&input.attrs);
        assert_eq!(
            doc,
            Some("This is a test function that does something useful.".to_string())
        );
    }

    #[test]
    fn test_validate_function_signature() {
        // Valid function
        let valid_fn: ItemFn = syn::parse_quote! {
            fn valid_tool(x: i32) -> String { "test".to_string() }
        };
        assert!(validate_function_signature(&valid_fn).is_ok());

        // Invalid function with self parameter
        let invalid_fn: ItemFn = syn::parse_quote! {
            fn invalid_tool(&self, x: i32) -> String { "test".to_string() }
        };
        assert!(validate_function_signature(&invalid_fn).is_err());

        // Invalid function with generics
        let generic_fn: ItemFn = syn::parse_quote! {
            fn generic_tool<T>(x: T) -> String { "test".to_string() }
        };
        assert!(validate_function_signature(&generic_fn).is_err());
    }

    #[test]
    fn test_parameter_count_limit() {
        // Create a function with exactly 50 parameters (should pass)
        let params_50 = (0..50)
            .map(|i| format!("p{}: i32", i))
            .collect::<Vec<_>>()
            .join(", ");
        let fn_with_50 = syn::parse_str::<ItemFn>(&format!(
            "fn tool_50({}) -> String {{ \"test\".to_string() }}",
            params_50
        ))
        .unwrap();
        let result_50 = tool_impl(TokenStream::new(), quote::quote! { #fn_with_50 });
        assert!(result_50.is_ok(), "50 parameters should be allowed");

        // Create a function with 51 parameters (should fail)
        let params_51 = (0..51)
            .map(|i| format!("p{}: i32", i))
            .collect::<Vec<_>>()
            .join(", ");
        let fn_with_51 = syn::parse_str::<ItemFn>(&format!(
            "fn tool_51({}) -> String {{ \"test\".to_string() }}",
            params_51
        ))
        .unwrap();
        let result_51 = tool_impl(TokenStream::new(), quote::quote! { #fn_with_51 });
        assert!(result_51.is_err(), "51 parameters should be rejected");

        if let Err(e) = result_51 {
            let error_msg = e.to_string();
            assert!(
                error_msg.contains("cannot have more than 50 parameters"),
                "Error message should mention the parameter limit, got: {}",
                error_msg
            );
        }
    }
}
