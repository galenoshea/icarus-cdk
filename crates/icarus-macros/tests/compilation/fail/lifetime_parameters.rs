use icarus_macros::tool;

/// This should fail - lifetime parameters are not supported
#[tool]
fn lifetime_tool<'a>(x: &'a str) -> &'a str {
    x
}

fn main() {}