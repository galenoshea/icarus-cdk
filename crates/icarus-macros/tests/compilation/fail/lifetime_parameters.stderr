error: Unsupported feature: Generic functions - Tool functions cannot have generic parameters
 --> tests/compilation/fail/lifetime_parameters.rs:5:17
  |
5 | fn lifetime_tool<'a>(x: &'a str) -> &'a str {
  |                 ^
